### PR TITLE
Improving PySolverInterface build

### PR DESCRIPTION
**Pull request by BenjaminRueth**
This resolves the following mentioned in #144. Additionally a smaller issue #68 has been resolved and tested with the dummy solver.

* where do we want to have the binaries? 
    * temporary in `precice/src/precice/bindings/python`
    * python module at default location (i.e. `site-packages` or `dist-packages`)
* update code influenced by binaries at python default locations 
    * python solverdummy
* compatibility
    * python 2 (2.7.12)
    * python 3 (3.5.2)

Issues on moving the build process of the python API into the main build process of preCICE are planned to be addressed in an independent step.

### CODE REVIEW

**Comment by BenjaminRueth** [comment]
There seems to be a possibility to add optional arguments to `setup.py`. This might help to allow defining a custom mpi compiler wrapper. See https://stackoverflow.com/questions/677577/distutils-how-to-pass-a-user-defined-parameter-to-setup-py for example.

**Comment by floli** [comment]
Why not give the users the possibility to define the MPI compiler wrapper and use that one? Either via an environment variable, or - better - using an argument to setup.py.

File: src/precice/bindings/python/setup.py [hunk 1]
```diff
@@ -1,62 +1,58 @@
...
 
-# check if PRECICE_ROOT is defined
-if not os.getenv('PRECICE_ROOT'):
-    print("ERROR: PRECICE_ROOT not defined!")
-    exit(1)
-
-precice_root = os.getenv('PRECICE_ROOT')
-python_bindings_path = os.path.join(precice_root,'src','precice','bindings','python')
+python_bindings_path = os.path.dirname(os.path.abspath(__file__))
+precice_root = os.path.join(python_bindings_path,"../../../../../precice")
```
Comment (floli) [inline src/precice/bindings/python/setup.py]: Why not $PRECICE_ROOT here?

It's ok, if you prefer the relative location, not sure if using PRECICE_ROOT is superior here.

Comment (BenjaminRueth) [reply src/precice/bindings/python/setup.py]: I used `$PRECICE_ROOT` before. But this gave me problems, when I had to run `sudo python setup.py install`, if the python is installed in a folder without user access.

**Comment by BenjaminRueth** [comment]
with d8e1923 you can now specify the mpi compiler wrapper through `--mpicompiler=` it should work with `setup.py build/install/build_ext`.

**Comment by uekerman** [comment]
why not `--compiler=`? then it would be consistent with scons, right?

**Comment by BenjaminRueth** [comment]
using `--compiler=` instead of `--mpicompiler=` is no problem. I will push the corresponding commit in a moment

**Comment by BenjaminRueth** [comment]
seems like I promised too much: If I use `--compiler=` instead of `--mpicompiler=`, I get strange error messages. The keyword `compiler` clashes with some other options and `mpicompiler` does not. `install` and `build` do not use the keywork `compiler`, but `build_ext` does. Therefore, I would like to stay with `--mpicompiler`.

**Comment by floli** [comment]
Looks good, with regard to compiling against non-default mpich:
```
% python setup.py build --mpicompiler=/opt/mpich/bin/mpicxx
running build
#####
calling my_build
using --mpicompiler=/opt/mpich/bin/mpicxx
adding extension
#####
running build_ext
#####
calling my_build_ext
using --mpicompiler=mpic++
#####
Compiling /home/lindnefn/precice/src/precice/bindings/python/PySolverInterface.pyx because it changed.
[1/1] Cythonizing /home/lindnefn/precice/src/precice/bindings/python/PySolverInterface.pyx
building 'PySolverInterface' extension
creating build
creating build/temp.linux-x86_64-3.7
creating build/temp.linux-x86_64-3.7/home
creating build/temp.linux-x86_64-3.7/home/lindnefn
creating build/temp.linux-x86_64-3.7/home/lindnefn/precice
creating build/temp.linux-x86_64-3.7/home/lindnefn/precice/src
creating build/temp.linux-x86_64-3.7/home/lindnefn/precice/src/precice
creating build/temp.linux-x86_64-3.7/home/lindnefn/precice/src/precice/bindings
creating build/temp.linux-x86_64-3.7/home/lindnefn/precice/src/precice/bindings/python
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -fPIC -I/home/lindnefn/precice/src/precice/bindings/python/../../../../../precice -I/usr/include/python3.7m -c /home/lindnefn/precice/src/precice/bindings/python/PySolverInterface.cpp -o build/temp.linux-x86_64-3.7/home/lindnefn/precice/src/precice/bindings/python/PySolverInterface.o -I/home/lindnefn/precice/src/precice/bindings/python/../../../../../precice -Wall -std=c++11 -g -I/opt/mpich/include -L/opt/mpich/lib -lmpicxx -Wl,-rpath -Wl,/opt/mpich/lib -Wl,--enable-new-dtags -lmpi
creating build/lib.linux-x86_64-3.7
g++ -pthread -shared -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now build/temp.linux-x86_64-3.7/home/lindnefn/precice/src/precice/bindings/python/PySolverInterface.o -L/usr/lib -lpython3.7m -o build/lib.linux-x86_64-3.7/PySolverInterface.cpython-37m-x86_64-linux-gnu.so -L/home/lindnefn/precice/src/precice/bindings/python/../../../../../precice/build/last/ -lprecice -g -I/opt/mpich/include -L/opt/mpich/lib -lmpicxx -Wl,-rpath -Wl,/opt/mpich/lib -Wl,--enable-new-dtags -lmpi
```

```
% ldd build/lib.linux-x86_64-3.7/PySolverInterface.cpython-37m-x86_64-linux-gnu.so | grep mpi
        libmpicxx.so.12 => /opt/mpich/lib/libmpicxx.so.12 (0x00007f99d503c000)
        libmpi.so.12 => /opt/mpich/lib/libmpi.so.12 (0x00007f99d4b8f000)
        libmpifort.so.12 => /opt/mpich/lib/libmpifort.so.12 (0x00007f99ce801000)
```

**Pull request merged**